File: examples/async/bitmart_ws_private_readonly.py
"""Open a BitMart private WebSocket user-data stream."""

import asyncio
import os

from dcex.ws import bitmart


async def main() -> None:
    """Open the private stream, subscribe to balance updates, and print one message."""
    api_key = os.environ["BITMART_API_KEY"]
    api_secret = os.environ["BITMART_API_SECRET"]
    memo = os.environ["BITMART_MEMO"]

    async with bitmart.private(api_key=api_key, api_secret=api_secret, memo=memo) as ws:
        print(await ws.recv())
        await ws.subscribe_balance()
        print(await ws.recv())


if __name__ == "__main__":
    asyncio.run(main())

File: crates/dcex/examples/bitmart_ws_private_readonly.rs
use std::env;
use std::time::Duration;

use dcex::ws::bitmart::BitmartPrivateWebSocket;

#[tokio::main]
async fn main() -> dcex::Result<()> {
    let api_key = env::var("BITMART_API_KEY")
        .map_err(|error| dcex::DcexError::InvalidInput(error.to_string()))?;
    let api_secret = env::var("BITMART_API_SECRET")
        .map_err(|error| dcex::DcexError::InvalidInput(error.to_string()))?;
    let memo = env::var("BITMART_MEMO")
        .map_err(|error| dcex::DcexError::InvalidInput(error.to_string()))?;

    let mut ws = BitmartPrivateWebSocket::new(api_key, api_secret, memo, Duration::from_secs(10))?;
    ws.connect().await?;
    println!("{}", ws.recv().await?);
    ws.subscribe_balance().await?;
    println!("{}", ws.recv().await?);
    ws.close().await?;
    Ok(())
}
